### ISSUE TITLE
Only warn about small files when they are not empty

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#1065]
This is a follow up to https://github.com/timberio/vector/pull/863#issuecomment-544314174.

To recap, when the `file` source is configured to use the `checksum` fingerprinting strategy a _warning_ log message is emitted letting the user know the file is too small to generate a fingerprint:

```rust
warn!(message = "Ignoring file smaller than fingerprint_bytes", file = ?path);
```

This is a very welcome feature for files that are actually too small. It helps to prevent confusion around why small files aren't being tailed. Unfortunately, it can produce a lot of noise in production environments where many empty files are present.

I'm unsure if we should ignore empty files, given that it could lead to the same confusion that we experienced previously with small files. Then again, I can see the reasoning behind ignoring empty files. I'm opening this up for discussion so we can come to a conclusion about what to do in this scenario.

**Comment by binarylogic** [comment]
@karlseguin do you still feel strongly about this? I'm on the fence, so I'm leaning towards closing this. If this is causing you a lot of pain/noise we can probably implement a quick fix.

**Comment by karlseguin** [comment]
Short:
I don't feel strongly about it, no. Feel free to close.

Long: 

Not related specifically to this issue, but in general I've found the file source with small files difficult to use. Either because of the above, or because it warns a lot about duplicate files. For example, we run Lynis daily and grep for warnings into a file. We might end up with something like:
```
lynis.error = "[warning] system needs a reboot"
lynis.error.1 = "[warning] system needs a reboot"
lynis.error.2 = "[warning] system needs a reboot"
```

I think at best this generates warnings, and at worse might lead to some messages being skipped (because it thinks it's duplicates).

I say i *think* that's what happens because we've largely solved this by doing one of three things, so I don't really remember the exact behavior. We do:

1 - Use the journal source where possible
2 - I've generate extra noise for these very short things (like adding the date to the start of the output)
3 - I've moved to device_and_inode for anything else

Even less related, but a bigger issue is: https://github.com/timberio/vector/issues/1391 which has forced me to just ignore vector logs altogether. Now, that's very lazy of me, as I could just use vector to filter out those messages. I might feel more strongly about questionable use of WARN and ERROR levels if I wasn't ignore vector's logs altogether.

**Comment by Nello-Angelo** [comment]
```
2023-07-24T22:13:28.614435Z  WARN source{component_kind="source" component_id=kube_logs component_type=kubernetes_logs component_name=kube_logs}:file_server: vector::internal_events::file::source: Currently ignoring file too small to fingerprint. file=/var/log/pods/longhorn-system_engine-image-ei-74783864-5wpcs_18b56f61-37a5-4982-b99a-69204c0848a2/engine-image-ei-74783864/2.log
2023-07-24T22:13:28.614605Z  WARN source{component_kind="source" component_id=kube_logs component_type=kubernetes_logs component_name=kube_logs}:file_server: vector::internal_events::file::source: Currently ignoring file too small to fingerprint. file=/var/log/pods/longhorn-system_engine-image-ei-74783864-5wpcs_18b56f61-37a5-4982-b99a-69204c0848a2/engine-image-ei-74783864/3.log
```

**Comment by argais** [comment]
I've landed here searching for this same error message, does this mean the file wont be captured at all?
Context on my side: I run around 15000 jobs a day in kubernetes clusters and some really quick jobs (run in about a second or so) do not have their logs captured at all and this is the only suspicious message in the vector logs.

**Comment by jszwedko** [comment]
> I've landed here searching for this same error message, does this mean the file wont be captured at all? Context on my side: I run around 15000 jobs a day in kubernetes clusters and some really quick jobs (run in about a second or so) do not have their logs captured at all and this is the only suspicious message in the vector logs.

What version of Vector? This log message is actually only valid for old Vector versions which read the first `n` bytes to fingerprint files, but newer versions of Vector read the first line of the file (by default) instead. Do the files in question have more than one line in them?

**Comment by mmerickel** [comment]
@jszwedko I see this log all the time right now on vector 0.40. Basically on any job that starts/ends quickly I think. It's unclear why because there's not more information and I can't catch it in time but it's a newer version vector going back to your comment. If it is an empty file or something I definitely would not like to see this at a warning level polluting the logs.

**Comment by jszwedko** [comment]
> @jszwedko I see this log all the time right now on vector 0.40. Basically on any job that starts/ends quickly I think. It's unclear why because there's not more information and I can't catch it in time but it's a newer version vector going back to your comment. If it is an empty file or something I definitely would not like to see this at a warning level polluting the logs.

Gotcha, yeah, I'm assuming you are seeing "Currently ignoring file too small to fingerprint.". My comment above was about the log line mentioned in the original issue ("Ignoring file smaller than fingerprint_bytes") which shouldn't be emitted anymore.

**Comment by mmerickel** [comment]
You're correct I did misread the original ticket. My comments were about the "Currently ignoring file too small to fingerprint". I can make a separate ticket if you prefer but it is definitely an irritation with the kubernetes_log source because I can't do anything about it and I can't tell if it's actually a problem. I do see logs from many of those pods eventually pulled into vector but obviously there's no message saying that - I only see the warning from vector. :-)

Separately from that, I was getting into this issue as part of trying to figure out why some logs are missed by vector completely on short-lived pods. I haven't quite found an open ticket for this but was looking - there is `glob_minimum_cooldown_ms` but none of that is super reliable to avoid missing log files on short-lived pods.

**Comment by jszwedko** [comment]
Yeah, for sure, I agree that is a nuisance. I think we can reuse this issue to cover "Currently ignoring file too small to fingerprint" since the original log line is irrelevant now.

**Comment by an-cor** [comment]
Hi! I’m interested in working on this issue.

From reading the thread, my understanding is that the current problem is that Vector can emit the “Currently ignoring file too small to fingerprint” warning even for empty files, which can create noisy logs in production environments. My plan is to first reproduce where this warning is emitted in the file source, then update the logic so empty files are handled without warning while non-empty files that are too small still warn.

Please let me know if that matches the intended direction.
